File: backend/src/utils/resume_parser.py
#!/usr/bin/env python3
"""
Resume Parser Utility
Extracts text content from PDF and DOCX resumes
"""

import sys
import json
import os
from pathlib import Path

def parse_pdf(file_path):
    """Parse PDF resume and extract text"""
    try:
        import PyPDF2
        with open(file_path, 'rb') as file:
            pdf_reader = PyPDF2.PdfReader(file)
            text = ""
            for page in pdf_reader.pages:
                text += page.extract_text() + "\n"
            return text.strip()
    except ImportError:
        return {"error": "PyPDF2 not installed. Run: pip install PyPDF2"}
    except Exception as e:
        return {"error": f"Error parsing PDF: {str(e)}"}

def parse_docx(file_path):
    """Parse DOCX resume and extract text"""
    try:
        from docx import Document
        doc = Document(file_path)
        text = ""
        for paragraph in doc.paragraphs:
            text += paragraph.text + "\n"
        return text.strip()
    except ImportError:
        return {"error": "python-docx not installed. Run: pip install python-docx"}
    except Exception as e:
        return {"error": f"Error parsing DOCX: {str(e)}"}

def parse_txt(file_path):
    """Parse plain text resume"""
    try:
        with open(file_path, 'r', encoding='utf-8') as file:
            return file.read().strip()
    except Exception as e:
        return {"error": f"Error reading text file: {str(e)}"}

def parse_resume(file_path):
    """Main function to parse resume based on file extension"""
    if not os.path.exists(file_path):
        return {"error": f"File not found: {file_path}"}
    
    file_extension = Path(file_path).suffix.lower()
    
    if file_extension == '.pdf':
        text = parse_pdf(file_path)
    elif file_extension in ['.docx', '.doc']:
        text = parse_docx(file_path)
    elif file_extension == '.txt':
        text = parse_txt(file_path)
    else:
        return {"error": f"Unsupported file format: {file_extension}"}
    
    if isinstance(text, dict) and "error" in text:
        return text
    
    # Extract basic information (simple extraction)
    lines = text.split('\n')
    
    return {
        "success": True,
        "full_text": text,
        "line_count": len(lines),
        "char_count": len(text),
        "preview": text[:500] if len(text) > 500 else text
    }

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print(json.dumps({"error": "Usage: python resume_parser.py <file_path>"}))
        sys.exit(1)
    
    file_path = sys.argv[1]
    result = parse_resume(file_path)
    print(json.dumps(result, indent=2))

File: backend/src/utils/resumeParser.ts
import fs from 'fs';
import path from 'path';
import * as pdfParse from 'pdf-parse';
import mammoth from 'mammoth';

interface ResumeParseResult {
  success: boolean;
  full_text?: string;
  line_count?: number;
  char_count?: number;
  preview?: string;
  error?: string;
}

/**
 * Parse PDF resume using pdf-parse
 * @param filePath - Absolute path to the PDF file
 * @returns Parsed text content
 */
async function parsePDF(filePath: string): Promise<string> {
  const dataBuffer = fs.readFileSync(filePath);
  // @ts-ignore - pdf-parse type definition issue
  const data = await pdfParse(dataBuffer);
  return data.text;
}

/**
 * Parse DOCX resume using mammoth
 * @param filePath - Absolute path to the DOCX file
 * @returns Parsed text content
 */
async function parseDOCX(filePath: string): Promise<string> {
  const result = await mammoth.extractRawText({ path: filePath });
  return result.value;
}

/**
 * Parse TXT resume
 * @param filePath - Absolute path to the TXT file
 * @returns Parsed text content
 */
async function parseTXT(filePath: string): Promise<string> {
  return fs.readFileSync(filePath, 'utf-8');
}

/**
 * Parse resume file using Node.js (no Python required)
 * @param filePath - Absolute path to the resume file
 * @returns Parsed resume content
 */
export async function parseResume(filePath: string): Promise<ResumeParseResult> {
  try {
    if (!fs.existsSync(filePath)) {
      return { success: false, error: `File not found: ${filePath}` };
    }

    const ext = path.extname(filePath).toLowerCase();
    let text = '';

    switch (ext) {
      case '.pdf':
        text = await parsePDF(filePath);
        break;
      case '.docx':
      case '.doc':
        text = await parseDOCX(filePath);
        break;
      case '.txt':
        text = await parseTXT(filePath);
        break;
      default:
        return { success: false, error: `Unsupported file format: ${ext}` };
    }

    if (!text || text.trim().length === 0) {
      return { success: false, error: 'No text content found in file' };
    }

    const lines = text.split('\n').filter((line) => line.trim());

    return {
      success: true,
      full_text: text,
      line_count: lines.length,
      char_count: text.length,
      preview: text.substring(0, 500),
    };
  } catch (error: any) {
    console.error('Resume parsing error:', error);
    return {
      success: false,
      error: error.message || 'Failed to parse resume',
    };
  }
}

/**
 * Generate interview questions based on resume content
 * @param resumeText - Extracted text from resume
 * @param interviewType - Type of interview (tech/hr)
 * @returns Prompt for AI model
 */
export function generateResumeBasedPrompt(
  resumeText: string,
  interviewType: 'tech' | 'hr',
  questionCount: number
): string {
  const resumePreview = resumeText.substring(0, 2000); // Limit to avoid token limits

  if (interviewType === 'tech') {
    return `Based on the following candidate's resume, generate ${questionCount} technical interview questions that are relevant to their experience, skills, and background. Focus on their technical skills, projects, and technologies mentioned.

Resume Content:
${resumePreview}

Generate ${questionCount} personalized technical questions that:
1. Test their knowledge of technologies they claim to know
2. Ask about specific projects they've worked on
3. Evaluate their problem-solving abilities based on their experience level
4. Are appropriate for their background and seniority

Return only the questions, one per line, numbered.`;
  } else {
    return `Based on the following candidate's resume, generate ${questionCount} HR interview questions that are relevant to their career history, experiences, and professional background.

Resume Content:
${resumePreview}

Generate ${questionCount} personalized HR questions that:
1. Explore their career transitions and choices
2. Ask about specific roles and responsibilities they've mentioned
3. Evaluate their achievements and leadership experiences
4. Assess cultural fit based on their background

Return only the questions, one per line, numbered.`;
  }
}
